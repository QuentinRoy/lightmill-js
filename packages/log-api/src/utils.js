function updateModelPropertiesInPlace(model, callback) {
  for (const [key, prop] of model.properties) {
    const updatedProp = callback(key, prop);
    model.properties.set(key, updatedProp);
  }
}

export function withPartial(_context, target, props) {
  let notFoundProps = new Set(props ?? []);
  updateModelPropertiesInPlace(target, (_key, prop) => {
    if (props == null || props.includes(prop.name)) {
      notFoundProps.delete(prop.name);
      return { ...prop, optional: true };
    }
    return prop;
  });
  if (notFoundProps.size > 0) {
    throw new Error(
      `Partial decorator: properties ${Array.from(notFoundProps).join(', ')} not found`,
    );
  }
}

export const $decorators = { Utils: { withPartial } };
