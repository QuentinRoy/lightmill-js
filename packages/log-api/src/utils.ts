import { getErrorSchema, getOneErrorDocumentSchema } from './jsonapi.ts';
import { z } from './zod-openapi.ts';

export function mapKeys<
  T extends object,
  M extends (k: Extract<keyof T, string>) => PropertyKey,
>(obj: T, fn: M) {
  return Object.fromEntries(
    Object.entries(obj).map(([key, value]) => {
      // @ts-expect-error: we assume T has no extra properties so key is
      // a valid key of T.
      return [fn(key), value];
    }),
  ) as { [K in keyof T as ReturnType<M>]: T[K] };
}

export function prefixKeys<T extends Record<string, unknown>, P extends string>(
  obj: T,
  prefix: P,
) {
  return mapKeys(obj, (k) => {
    return `${prefix}${k as keyof T & (string | number)}` as const;
  });
}

// Common types
export const StringOrArrayOfStrings = z
  .union([z.string(), z.array(z.string())])
  .openapi('StringOrArrayOfStrings');

export const ForbiddenErrorResponse = getOneErrorDocumentSchema(
  getErrorSchema({
    code: 'FORBIDDEN',
    statusText: 'Forbidden',
    statusCode: 403,
  }),
).openapi('ForbiddenErrorResponse');
