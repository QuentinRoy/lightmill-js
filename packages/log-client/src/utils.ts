import { JsonValue } from 'type-fest';

type JsonFetchUrl = string | URL;
type JsonFetchOptions<AllowBody extends boolean = true> = Omit<
  NonNullable<Parameters<typeof fetch>[1]>,
  'body' | 'method' | 'headers'
> & {
  body?: AllowBody extends true ? JsonValue : never;
  headers?: Record<string, string>;
};

function makeJsonFetchMethod(
  method: string,
  opt: { noBody: true },
): (url: JsonFetchUrl, options: JsonFetchOptions<false>) => Promise<unknown>;
function makeJsonFetchMethod(
  method: string,
  opt?: { noBody: false },
): (url: JsonFetchUrl, options: JsonFetchOptions) => Promise<unknown>;
function makeJsonFetchMethod(method: string, { noBody = false } = {}) {
  return async function fetchMethod(
    url: JsonFetchUrl,
    options: JsonFetchOptions,
  ) {
    let body = noBody ? undefined : JSON.stringify(options.body);
    let headers: JsonFetchOptions['headers'] = {
      ...options.headers,
      Accept: 'application/json',
    };
    if (body != null) {
      headers['Content-Type'] = 'application/json';
    }
    let response = await fetch(url, { ...options, method, body, headers });
    let json = await response.json();
    if (response.ok) {
      return json as unknown;
    } else {
      throw new Error(json.message ?? 'Unknown error');
    }
  };
}

export const post = makeJsonFetchMethod('POST');
export const put = makeJsonFetchMethod('PUT');
export const get = makeJsonFetchMethod('GET', { noBody: true });
export const del = makeJsonFetchMethod('DELETE');
export const patch = makeJsonFetchMethod('PATCH');
