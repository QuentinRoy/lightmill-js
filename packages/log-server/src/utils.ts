import { mapKeys } from 'remeda';
import { SnakeCase } from 'type-fest';
import { snakeCase } from 'change-case';
import { z } from 'zod';

export function arrayify<T>(value: T | T[], isEmptyWithUndefined?: false): T[];
export function arrayify<T>(
  value: T | T[] | undefined,
  isEmptyWithUndefined: true,
): T[];
export function arrayify<T>(value: T | T[], isEmptyWithUndefined = false): T[] {
  if (value === undefined && isEmptyWithUndefined) return [];
  return Array.isArray(value) ? value : [value];
}

export const JsonPrimitive = z.union([
  z.string(),
  z.number(),
  z.boolean(),
  z.null(),
]);
export type JsonPrimitive = z.infer<typeof JsonPrimitive>;
export type JsonValue =
  | JsonPrimitive
  | ({ [Key in string]: JsonValue } & {
      [Key in string]?: JsonValue | undefined;
    })
  | JsonValue[]
  | readonly JsonValue[];
export const JsonValue: z.ZodType<JsonValue> = z.lazy(() =>
  z.union([JsonPrimitive, z.array(JsonValue), z.record(JsonValue)]),
);
export const JsonObject = z.record(JsonValue);
export type JsonObject = z.infer<typeof JsonObject>;
export const JsonArray = z.array(JsonValue);
export type JsonArray = z.infer<typeof JsonArray>;

export function toSnakeCase<
  R extends Record<string | number | symbol, unknown>,
>(input: R) {
  return mapKeys(input, (key) =>
    typeof key === 'string' ? snakeCase(key) : key,
  ) as unknown as {
    [K in keyof R as SnakeCase<K>]: R[K];
  };
}
