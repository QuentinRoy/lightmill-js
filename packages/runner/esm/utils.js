// eslint-disable-next-line import/prefer-default-export
export const asyncReduce = (iterator, reducer, init) => {
  const next = () => Promise.resolve().then(() => iterator.next());

  const startup =
    init === undefined
      ? next().then(r => ({ initAcc: r.value, initIndex: 1 }))
      : Promise.resolve({ initIndex: 0, initAcc: init });

  const reduceEnd = ({ initAcc, initIndex }) =>
    next().then(
      ({ done, value }) =>
        done
          ? initAcc
          : Promise.resolve()
              .then(() => reducer(initAcc, value, initIndex))
              .then(acc =>
                reduceEnd({ initAcc: acc, initIndex: initIndex + 1 })
              )
    );

  return startup.then(reduceEnd);
};

export const asyncForEach = (iterator, callback) =>
  asyncReduce(
    iterator,
    (_, value, index) =>
      Promise.resolve()
        .then(() => callback(value, index))
        .then(() => {}),
    null
  );
