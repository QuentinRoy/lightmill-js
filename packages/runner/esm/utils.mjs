// Runs promises one after the other, and returns the results of all.
const pSeries = (promiseGetters) =>
  promiseGetters.reduce(
    (ps, p) => ps.then((psRes) => p().then((pRes) => [...psRes, pRes])),
    Promise.resolve([])
  );

// Runs promises in parallel, and returns the promise of all (like Promise.all
// but using promise getters so that it has the same signature than pSeries).
const pPar = (promiseGetters) => Promise.all(promiseGetters.map((p) => p()));

// Creates a promise from a value or another promise.
const pFrom = (v) => Promise.resolve().then(() => v);

// Make a function asynchronous by creating a promise from its value
// (the function calls remain synchronous though).
const asyncF =
  (f) =>
  (...args) =>
    pFrom(f(...args));

// Reduces an async iterator (like Array#reduce)
export const asyncReduce = (iterator, reducer, init, parallel = false) => {
  const sequencer = parallel ? pPar : pSeries;
  const aNext = asyncF((...args) => iterator.next(...args));
  const aReducer = asyncF(reducer);

  const reduceEnd = (result, { done, value }) =>
    done
      ? result
      : sequencer([() => aReducer(result, value), () => aNext()]).then((args) =>
          reduceEnd(...args)
        );

  return pFrom(init)
    .then((init_) =>
      init === undefined ? aNext() : { done: false, value: init_ }
    )
    .then(({ done, value }) =>
      done ? value : aNext().then((next) => reduceEnd(value, next))
    );
};

// Calls a callback for values from an async iterator. If the callback returns
// a promise, wait for it to be resolved before moving on.
export const asyncForEach = (iterator, callback) => {
  const aCallback = asyncF(callback);
  return asyncReduce(
    iterator,
    (_, value) => aCallback(value).then(() => {}),
    null
  );
};
