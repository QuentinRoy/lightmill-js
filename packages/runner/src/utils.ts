// Creates a promise from a value or another promise.
function pFrom<T>(v: T | PromiseLike<T>): Promise<T> {
  return Promise.resolve().then(() => v);
}

// Make a function asynchronous by creating a promise from its value
// (the function calls remain synchronous though).
function makeAsync<Args extends unknown[], R>(
  f: (...args: Args) => R | PromiseLike<R>
): (...args: Args) => Promise<R> {
  return (...args: Args) => pFrom(f(...args));
}

function isPromiseLike(p: unknown): p is PromiseLike<unknown> {
  return (
    p !== null &&
    typeof p === 'object' &&
    'then' in p &&
    typeof p.then === 'function'
  );
}

// Calls a callback for values from an async iterator. If the callback returns
// a promise, wait for it to be resolved before moving on.
export function asyncForEach<Item>(
  iterator: Iterator<Item> | AsyncIterator<Item>,
  callback: (value: Item, index: number) => PromiseLike<unknown>,
  forceAsync?: boolean
): Promise<void>;
export function asyncForEach<Item>(
  iterator: Iterator<Item>,
  callback: (value: Item, index: number) => unknown,
  forceAsync?: false
): void;
export function asyncForEach<Item>(
  iterator: Iterator<Item>,
  callback: (value: Item, index: number) => unknown,
  forceAsync: true
): Promise<void>;
export function asyncForEach<Item>(
  iterator: AsyncIterator<Item>,
  callback: (value: Item, index: number) => unknown,
  forceAsync?: boolean
): Promise<void>;
export function asyncForEach<Item>(
  iterator: Iterator<Item> | AsyncIterator<Item>,
  callback: (value: Item, index: number) => unknown | PromiseLike<unknown>,
  forceAsync?: boolean
): Promise<void> | void;
export function asyncForEach<Item>(
  iterator: Iterator<Item> | AsyncIterator<Item>,
  callback: (value: Item, index: number) => unknown | PromiseLike<unknown>,
  forceAsync = false
): Promise<void> | void {
  callback = forceAsync ? makeAsync(callback) : callback;
  function fetchNext(index: number): PromiseLike<void> | void {
    const next = forceAsync ? pFrom(iterator.next()) : iterator.next();
    if (isPromiseLike(next)) {
      return next.then((result) => onNext(result, index));
    } else {
      return onNext(next, index);
    }
  }
  function onNext(
    result: IteratorResult<Item>,
    index: number
  ): PromiseLike<void> | void {
    if (result.done) {
      return;
    }
    const callbackResult = callback(result.value, index);
    if (isPromiseLike(callbackResult)) {
      return callbackResult.then(() => fetchNext(index + 1));
    } else {
      return fetchNext(index + 1);
    }
  }
  let result = fetchNext(0);
  if (isPromiseLike(result)) {
    // Ensures we return a promise and not only a promise like.
    return pFrom(result).then(() => {
      return;
    });
  }
}
