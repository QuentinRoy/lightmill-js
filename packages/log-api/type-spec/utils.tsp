import "@typespec/http";
import "./json-api.tsp";
import "./utils.js";

using TypeSpec.Http;
using JsonApi;

namespace Utils;

extern dec withPartial(target: Reflection.Model, props?: valueof Array<string>);

model InvalidQueryParameterDoc
  is ErrorDocument<Error<"INVALID_QUERY_PARAMETER", HttpStatus.`400`> & {
    source: {
      parameter: string;
    };
  }>;

/**
 * Error document sent when an operation is rejected due to insufficient
 * permissions.
 */
model ForbiddenErrorDocTemplate<Code extends string = "FORBIDDEN">
  is ErrorDocument<Error<Code, HttpStatus.`403`>>;
model ForbiddenErrorDoc is ForbiddenErrorDocTemplate<"FORBIDDEN">;
model ForbiddenErrorResponse extends ForbiddenResponse {
  @body doc: ForbiddenErrorDoc;
  @header contentType: JsonApi.APIContentType;
}

model InvalidCredentialsError is Error<"INVALID_CREDENTIALS", HttpStatus.`403`>;
model InvalidCredentialsErrorResponse extends ForbiddenResponse {
  @body doc: ErrorDocument<InvalidCredentialsError>;
  @header contentType: JsonApi.APIContentType;
}

model NotFoundError<Code extends string = "NOT_FOUND">
  is Error<Code, HttpStatus.`404`>;
/**
 * Error document sent when a resource is not found.
 */
model NotFoundErrorDocTemplate<Code extends string = "NOT_FOUND">
  is ErrorDocument<NotFoundError<Code>>;
model NotFoundErrorDoc is NotFoundErrorDocTemplate<"NOT_FOUND">;
model NotFoundErrorResponse extends NotFoundResponse {
  @body doc: NotFoundErrorDoc;
  @header contentType: JsonApi.APIContentType;
}

/**
 * Error document sent when a request is made with an invalid session.
 */
model SessionRequiredError is Error<"SESSION_REQUIRED", HttpStatus.`403`>;
model SessionRequiredErrorResponse extends ForbiddenResponse {
  @body doc: ErrorDocument<SessionRequiredError>;
  @header contentType: JsonApi.APIContentType;
}

model UnsupportedMediaTypeError
  is Error<"UNSUPPORTED_MEDIA_TYPE", HttpStatus.`415`>;
/**
 * Error document sent when an unsupported media type is requested.
 */
model UnsupportedMediaTypeErrorDoc
  is ErrorDocument<Error<"UNSUPPORTED_MEDIA_TYPE", HttpStatus.`415`>>;
model UnsupportedMediaTypeErrorResponse extends Response<415> {
  @body doc: ErrorDocument<UnsupportedMediaTypeError>;
  @header contentType: JsonApi.APIContentType;
}

alias CookieSessionName = "lightmill-session-id";

// I don't want this to be in main because I don't want this model to be
// included in the generated OpenAPI schemas.
@doc("Cookie-based session")
model CookieSessionAuth {
  @doc("Http authentication")
  type: AuthType.apiKey;

  @doc("location of the API key")
  in: ApiKeyLocation.cookie;

  @doc("name of the API key")
  name: CookieSessionName;
}

@withPartial
model Partial<T> {
  ...T;
}
