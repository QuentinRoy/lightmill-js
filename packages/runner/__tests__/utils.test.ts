import { describe, expect, it, vi } from 'vitest';
import { asyncForEach } from '../src/utils.js';

function wait(t = 0) {
  return new Promise((resolve) => {
    setTimeout(resolve, t);
  });
}

const generators = {
  sync: () => {
    const arr = ['a', 'b', 'c', 'd'];
    return arr[Symbol.iterator]();
  },
  async: async function* async() {
    yield 'a';
    await wait(0);
    yield 'b';
    await wait(0);
    yield 'c';
    await wait(0);
    yield 'd';
  },
};
const throwingGenerators = {
  sync: function* syncThatThrows() {
    yield 'a';
    yield 'b';
    throw new Error('mock-generator-error');
  },
  async: async function* syncThatThrows() {
    yield 'a';
    await wait(0);
    yield 'b';
    await wait(0);
    throw new Error('mock-generator-error');
  },
};

type Deffered<V> = {
  promise: Promise<V>;
  resolve: () => void;
  reject: (err?: Error) => void;
};
function deffer<V>(value: V): Deffered<V>;
function deffer(): Deffered<void>;
function deffer<V>(value?: V) {
  let resolve: () => void;
  let reject: (err?: Error) => void;
  const promise = new Promise<V>((res, rej) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    resolve = () => res(value as any);
    reject = rej;
  });
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return { promise, resolve: resolve!, reject: reject! };
}

const describeEachSyncs = describe.each(['sync', 'async']);

describeEachSyncs(`asyncForEach with %s iterator`, (syncArg) => {
  const sync = syncArg as 'sync' | 'async';
  it('calls its callback for each values', async () => {
    const gen = generators[sync as 'sync' | 'async'];
    const callback = vi.fn();
    await asyncForEach(gen(), callback);
    expect(callback.mock.calls).toEqual([
      ['a', 0],
      ['b', 1],
      ['c', 2],
      ['d', 3],
    ]);
  });

  it('waits for its callback to resolve if it returned a promise', async () => {
    const gen = generators[sync];
    const defers = { b: deffer(), d: deffer() };
    const log: string[] = [];
    const callback = vi.fn((value) => {
      if (value in defers) {
        log.push(`${value}-async-start`);
        return defers[value].promise.then(() => {
          log.push(`${value}-async-end`);
        });
      } else {
        log.push(`${value}-sync`);
      }
    });
    const end = vi.fn();
    let forEachProm = (asyncForEach(gen(), callback) as Promise<void>).then(
      () => end()
    );
    expect(end.mock.calls).toEqual([]);
    await wait(5);
    expect(log).toEqual(['a-sync', 'b-async-start']);
    defers.b.resolve();
    await wait(5);
    expect(end.mock.calls).toEqual([]);
    expect(log).toEqual([
      'a-sync',
      'b-async-start',
      'b-async-end',
      'c-sync',
      'd-async-start',
    ]);
    expect(end.mock.calls).toEqual([]);
    defers.d.resolve();
    await forEachProm;
    expect(log).toEqual([
      'a-sync',
      'b-async-start',
      'b-async-end',
      'c-sync',
      'd-async-start',
      'd-async-end',
    ]);
    expect(end.mock.calls).toEqual([[]]);
  });

  it('rejects if a sync callback throws', async () => {
    const gen = generators[sync as 'sync' | 'async'];
    const callback = vi.fn((v) => {
      if (v === 'b') throw new Error('mock-callback-error');
    });
    if (sync === 'sync') {
      expect(() => asyncForEach(gen(), callback)).toThrow(
        'mock-callback-error'
      );
    } else {
      await expect(asyncForEach(gen(), callback)).rejects.toThrow(
        'mock-callback-error'
      );
    }
    expect(callback.mock.calls).toEqual([
      ['a', 0],
      ['b', 1],
    ]);
  });

  it('rejects if an async callback throws', async () => {
    const gen = generators[sync as 'sync' | 'async'];
    const callback = vi.fn((v) =>
      wait(0).then(() => {
        if (v === 'b') throw new Error('mock-callback-error');
      })
    );
    await expect(asyncForEach(gen(), callback)).rejects.toThrow(
      'mock-callback-error'
    );
    expect(callback.mock.calls).toEqual([
      ['a', 0],
      ['b', 1],
    ]);
  });

  it('rejects if the iterator throws', async () => {
    const gen = throwingGenerators[sync as 'sync' | 'async'];
    const callback = vi.fn();
    if (sync === 'sync') {
      expect(() => asyncForEach(gen(), callback)).toThrow(
        'mock-generator-error'
      );
    } else {
      await expect(asyncForEach(gen(), callback)).rejects.toThrow(
        'mock-generator-error'
      );
    }
    expect(callback.mock.calls).toEqual([
      ['a', 0],
      ['b', 1],
    ]);
  });
});
