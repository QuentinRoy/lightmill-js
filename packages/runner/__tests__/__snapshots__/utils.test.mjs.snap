// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asyncForEach with async iterator rejects if a sync callback throws 1`] = `
[MockFunction] {
  "calls": [
    [
      "a",
    ],
    [
      "b",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "throw",
      "value": [Error: mock-callback-error],
    },
  ],
}
`;

exports[`asyncForEach with async iterator rejects if an async callback throws 1`] = `
[MockFunction] {
  "calls": [
    [
      "a",
    ],
    [
      "b",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`asyncForEach with async iterator rejects if the iterator throws 1`] = `
[MockFunction] {
  "calls": [
    [
      "a",
      "b",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`asyncForEach with sync iterator rejects if a sync callback throws 1`] = `
[MockFunction] {
  "calls": [
    [
      "a",
    ],
    [
      "b",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "throw",
      "value": [Error: mock-callback-error],
    },
  ],
}
`;

exports[`asyncForEach with sync iterator rejects if an async callback throws 1`] = `
[MockFunction] {
  "calls": [
    [
      "a",
    ],
    [
      "b",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`asyncForEach with sync iterator rejects if the iterator throws 1`] = `
[MockFunction] {
  "calls": [
    [
      "a",
      "b",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`asyncReduce with async iterator rejects if a sync reducers throws 1`] = `
[MockFunction] {
  "calls": [
    [
      "test:",
      "a",
    ],
    [
      "test:a,",
      "b",
    ],
    [
      "test:a,b,",
      "c",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": "test:a,",
    },
    {
      "type": "return",
      "value": "test:a,b,",
    },
    {
      "type": "throw",
      "value": [Error: mock-reducer-error],
    },
  ],
}
`;

exports[`asyncReduce with async iterator rejects if an async reducers throws 1`] = `
[MockFunction] {
  "calls": [
    [
      "test:",
      "a",
    ],
    [
      "test:a,",
      "b",
    ],
    [
      "test:a,b,",
      "c",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`asyncReduce with async iterator rejects if the iterator throws 1`] = `
[MockFunction] {
  "calls": [
    [
      "test:",
      "a",
    ],
    [
      "test:a,",
      "b",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": "test:a,",
    },
    {
      "type": "return",
      "value": "test:a,b,",
    },
  ],
}
`;

exports[`asyncReduce with sync iterator rejects if a sync reducers throws 1`] = `
[MockFunction] {
  "calls": [
    [
      "test:",
      "a",
    ],
    [
      "test:a,",
      "b",
    ],
    [
      "test:a,b,",
      "c",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": "test:a,",
    },
    {
      "type": "return",
      "value": "test:a,b,",
    },
    {
      "type": "throw",
      "value": [Error: mock-reducer-error],
    },
  ],
}
`;

exports[`asyncReduce with sync iterator rejects if an async reducers throws 1`] = `
[MockFunction] {
  "calls": [
    [
      "test:",
      "a",
    ],
    [
      "test:a,",
      "b",
    ],
    [
      "test:a,b,",
      "c",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`asyncReduce with sync iterator rejects if the iterator throws 1`] = `
[MockFunction] {
  "calls": [
    [
      "test:",
      "a",
    ],
    [
      "test:a,",
      "b",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": "test:a,",
    },
    {
      "type": "return",
      "value": "test:a,b,",
    },
  ],
}
`;
